### PR TITLE
Condition upgrades

### PR DESCRIPTION
**Pull request by abiosoft**
* Makes `if` condition reusable.
* Adds `if` condition to `redir`
* Adds `if_op` config to specify `and` or `or`. Defaults to `and` if omitted.

#### Example
Or is now possible. This will be `or`ed instead of `and`ed.
```
rewrite {
    if       {path} is /home
    if       {path} is /blog
    if_op    or
}
```
If condition is now available to `redir`. [This](https://forum.caddyserver.com/t/redirect-based-on-header-value-from-load-balancer/165/5?u=abiosoft) is now possible.
```
redir {
    if  {>X-Forwarded-Proto} is http
    /   https://{uri}
}
```
Any middleware/plugin can now integrate and support `if` condition using `httpserver.SetupIfMatcher` as it is now done with `redir` and `rewrite`.

### CODE REVIEW

File: caddyhttp/httpserver/condition.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+//        continue
+//      }
+//    }
+//  }
+//  ...
+//  // Check conditions in middleware/plugin httpserver.Handler.
+//  func (m Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+//    for _, c := range m.Configs { if c.Match(r) { ... } ... }
+//  }
+//
```
Comment (mholt) [inline caddyhttp/httpserver/condition.go]: Are these comments needed for this commit?

**Comment by mholt** [comment]
It looks like `if_type` (from your comment) is actually `if_cond` (in the code) - which do you think is better?

File: caddyhttp/rewrite/rewrite.go [hunk 2]
```diff
@@ -230,6 +228,10 @@ func (r *ComplexRule) matchExt(rPath string) bool {
 	return true
 }
 
+func newReplacer(r *http.Request) httpserver.Replacer {
+	return httpserver.NewReplacer(r, nil, "")
+}
+
```
Comment (mholt) [inline caddyhttp/rewrite/rewrite.go]: I think this is only used once (at least, that's what a quick Cmd+F on this page tells me) - maybe not needed then?

Comment (abiosoft) [inline caddyhttp/rewrite/rewrite.go]: It is also used in tests. No real benefits other being shorter.

File: caddyhttp/httpserver/condition.go (hunk 1 above)
Comment (abiosoft) [inline caddyhttp/httpserver/condition.go]: Not really. Should be removed in favour of a wiki entry.

**Comment by abiosoft** [comment]
> It looks like `if_type` (from your comment) is actually `if_cond` (in the code) - which do you think is better?

Apparently, proof that I'm human 😄. I think I vote for `if_type`.

**Comment by abiosoft** [comment]
How about `if_logic` ?

**Comment by mholt** [comment]
`if_logic` is good! Let's do that one.

File: caddyhttp/httpserver/condition.go (hunk 1 above)
Comment (mholt) [inline caddyhttp/httpserver/condition.go]: I agree. Please do that so we can eliminate this TODO.

**Comment by mholt** [comment]
Thanks Abiola! This is great. :smile:  :tada:

**Pull request merged**